File: BlinkyCloud/CPP/CppBlinkyStartup/StartupTask/BackgroundTask.cpp
 /********************************************************
*                                                        *
*   © Copyright (C) Microsoft. All rights reserved.      *
*                                                        *
*********************************************************/


#include "pch.h"
#include "BackgroundTask.h"

using namespace Windows::Foundation;
using namespace Windows::Devices::Enumeration;
using namespace Windows::Devices::Gpio;
using namespace concurrency;

using namespace StartupTask;
using namespace Platform;

using namespace Microsoft::IoT::Maker;

BackgroundTask::BackgroundTask()
{
}

void BackgroundTask::Run(IBackgroundTaskInstance^ taskInstance)
{
    svc = ref new MakerClient();
    Deferral = taskInstance->GetDeferral();

    TimerElapsedHandler ^handler = ref new TimerElapsedHandler(
        [this](ThreadPoolTimer ^timer) 
        {
            if (LEDStatus == 0)
            {
                LEDStatus = 1;
                if (pin != nullptr)
                    pin->Write(GpioPinValue::High);
            }
            else
            {
                LEDStatus = 0;
                if (pin != nullptr)
                    pin->Write(GpioPinValue::Low);
            }
            svc->SendData("LEDStatus", LEDStatus);    // Upload data to cloud
        }
    );
    TimeSpan interval;
    interval.Duration = 1000 * 1000 * 10;
    Timer = ThreadPoolTimer::CreatePeriodicTimer(handler, interval);

    InitGpio();

    svc->CommandReceived += ref new CommandReceivedEventHandler(
        [this, handler, interval](Object^ sender, CommandReceivedEventArgs^ e)
        {
            // handle commands from cloud
            if (e->CommandName == L"BlinkOn")
            {
                if (Timer == nullptr)
                {
                    Timer = ThreadPoolTimer::CreatePeriodicTimer(handler, interval);
                }
            }
            else if (e->CommandName == L"BlinkOff")
            {
                if (Timer != nullptr)
                    Timer->Cancel();
                Timer = nullptr;
            }
        }
    );

    svc->SendEvent(L"BlinkyStart", L"");        // Upload event to cloud
}

void BackgroundTask::InitGpio()
{
    create_task(GpioController::GetDefaultAsync()).then([this](task<GpioController ^> controllerOp) {
        auto gpio = controllerOp.get();

        if (gpio == nullptr)
        {
            pin = nullptr;
            return;
        }

        pin = gpio->OpenPin(LED_PIN);

        if (pin == nullptr)
        {
            return;
        }

        pin->Write(GpioPinValue::High);
        pin->SetDriveMode(GpioPinDriveMode::Output);
    });
}

File: BlinkyCloud/CS/CsBlinkyStartup/StartupTask/BackgroundTask.cs
// © Copyright (C) Microsoft. All rights reserved.

using System;
using Windows.ApplicationModel.Background;
using Windows.Devices.Gpio;
using Windows.System.Threading;
using System.Diagnostics;
using Microsoft.IoT.Maker;

namespace StartupTask
{
    public sealed class BackgroundTask : IBackgroundTask
    {
        MakerClient svc = new MakerClient();
        BackgroundTaskDeferral _deferral;
        public void Run(IBackgroundTaskInstance taskInstance)
        {
            _deferral = taskInstance.GetDeferral();
            timer = ThreadPoolTimer.CreatePeriodicTimer(Timer_Tick, TimeSpan.FromMilliseconds(1000));
            InitGPIO();

            svc.CommandReceived += OnCommand;      // Handle commands from cloud

            svc.SendEvent("BlinkyStart", "");      // Send event to cloud
        }

        private async void InitGPIO()
        {
            var gpio = await GpioController.GetDefaultAsync();

            if (gpio == null)
            {
                pin = null;
                return;
            }

            pin = gpio.OpenPin(LED_PIN);

            if (pin == null)
            {
                return;
            }

            pin.Write(GpioPinValue.High);
            pin.SetDriveMode(GpioPinDriveMode.Output);
        }

        private void Timer_Tick(ThreadPoolTimer timer)
        {
            if (LEDStatus == 0)
            {
                LEDStatus = 1;
                if (pin != null)
                    pin.Write(GpioPinValue.High);
            }
            else
            {
                LEDStatus = 0;
                if (pin != null)
                    pin.Write(GpioPinValue.Low);
            }
            Debug.WriteLine("LEDStatus=" + LEDStatus);

            svc.SendData("LEDStatus", LEDStatus);    // Upload LEDStatus to cloud
        }

        private void OnCommand(object sender, CommandReceivedEventArgs e)
        {
            switch (e.CommandName)
            {
                case "BlinkOn":
                    if (timer == null)
                        timer = ThreadPoolTimer.CreatePeriodicTimer(Timer_Tick, TimeSpan.FromMilliseconds(1000));
                    break;
                case "BlinkOff":
                    if (timer != null)
                        timer.Cancel();
                    timer = null;
                    break;
            }
        }

        private int LEDStatus = 0;
        private const int LED_PIN = 0;
        private GpioPin pin;
        private ThreadPoolTimer timer = null;
    }
}
